techmatters/hrm
# CHI-2825: Streaming AI dataset generator

## Description

Adds a new HRM task. We may not keep it around permanently. It is to generate training sets for predicting the category of a contact.

We should delete this task in the future once it's usefulness is past

Process:

1. Query the DB for contacts with transcript locations and categories
2. Stream the records from the query to a pipeline that looks up the transcript in S3 based on the location retrieved in the db, then adds the text to the object in the stream
3. Serializes each object in the stream as JSON and saves it as a separate s3 document
4. Streams the same JSON to a single file where it is stored as a line separated list of JSONs (as opposed to an array)

TODO:

Add permission to write to the 'ai' buckets to the {environment}-{region}-s3-data-policy in Terraform - I temporarily did this for development in the console: https://us-east-1.console.aws.amazon.com/iam/home?region=us-east-1#/policies/details/arn%3Aaws%3Aiam%3A%3A712893914485%3Apolicy%2Fdevelopment-us-east-1-s3-data-policy?section=permissions

### Checklist

### Other Related Issues

None

### Verification steps

To run the importer against dev transcripts:

1. Ensure the code is deployed to development using the Deploy HRM to Amazon ECS workflow
2. Go to the ECS console in AWS console, then open development-ecs-cluster -> tasks (tab at bottom)
3. Select 'Run New Task'
4. In the form 'Deployment configuration' section - ensure 'development-hrm-scheduled-task' is selected under 'Family'
5. In the form 'Networking' section ensure the `development-vpc` VPC is selected. Also ensure the `default`, `development-rds-src-sg`, `development-scheduled-task-sg` and `development-ecs-tasks-sg` security groups are selected (these won't be options if the development  VPC hasn't been selected)
6. In the form `Task Overrides` ensure 'Task Execution Role' is set to `development-ecsTaskExecutionRole`
7. In the form 'Container overrides' section, under transcript-scrubber ensure command override is set to `npm,run,start:generate-ai-training-set,development,AS;E2E,tl-aselo-ai-development` and AWS_REGION -> `us-east-1` is added to the environment variables
8. Click 'Create'
9. You should see the new task in the list of tasks 'Provisioning' - you can open it and review the logs

To check the output, look at the `tl-aselo-ai-development` contents. It should have one large file under the root of each short code, then a timestamped folder with the separate files per contact inside.

You can modify the above instructions to use staging (remember to change the shortcode in the command override) - but you will need to update the data policy for staging (see in description above)

To run the importer against production transcripts FROM A DEVELOPMENT DEPLOY (I already temporarily updated the s3 policy for prod the same way I did for dev):

1. Ensure the code is deployed to development using the Deploy HRM to Amazon ECS workflow
2. Go to the ECS console in AWS console, then open development-ecs-cluster -> tasks (tab at bottom)
3. Select 'Run New Task'
4. In the form 'Deployment configuration' section - ensure 'development-hrm-scheduled-task' is selected under 'Family'
5. In the form 'Networking' section ensure the `production-vpc` VPC is selected. Also ensure the `default`, `production-rds-src-sg`, `development-scheduled-task-sg` and `development-ecs-tasks-sg` security groups are selected (these won't be options if the development  VPC hasn't been selected)
6. In the form `Task Overrides` ensure 'Task Role' is overridden to `production-ecsTaskRole` 'Task Execution Role' is set to `production-ecsTaskExecutionRole`
7. In the form 'Container overrides' section, under transcript-scrubber ensure command override is set to `npm,run,start:generate-ai-training-set,production,AS,tl-aselo-ai-production` and AWS_REGION -> `us-east-1` is added to the environment variables. Ensure the file listed in 'Environment Files' is changed from `arn:aws:s3:::tl-hrm-vars-development/development.env` to `arn:aws:s3:::tl-hrm-vars-production/production.env`.
8. Click 'Create'
9. You should see the new task in the list of tasks 'Provisioning' - you can open it and review the logs

This will use the current AS Production transcripts. To use the scrubbed transcripts Katy saved for us, use the command `npm,run,start:generate-ai-training-set,production,JM;ZM,tl-aselo-ai-production-us-east-1,tl-aselo-ai-production-us-east-1`

### AFTER YOU MERGE

1. Cut a release tag using the GitHub workflow. Wait for it to complete and notify in the #aselo-deploys Slack channel.
10. Comment on the ticket with the release tag version AND any additional instructions required to configure an environment to test the changes.
11. Only then move the ticket into the QA column in JIRA

You are responsible for ensuring the above steps are completed. If you move a ticket into QA without advising what version to test, the QA team will assume the latest tag has the changes. If it does not, the following confusion is on you! :-P
